fish_completions.go

```diff
@@ -89,6 +89,33 @@ function __%[1]s_perform_completion
     printf "%%s\n" "$directiveLine"
 end
 
+function __%[1]s_doesnt_requires_order_preservation
+    __%[1]s_debug ""
+    __%[1]s_debug "========= checking if order preservation is required =========="
+
+    set -l results (__%[1]s_perform_completion)
```

Comment:
With the new keeporder logic `__%[1]s_perform_completion` ends up being called 3 times instead of one.  That function calls the program to get the completions, so there is some expense to that call.

I haven't thought it through yet, but it would be nice if we could achieve all the logic by using only a single call to the program.

Reply:
I am not super familiar with fish, a work around I could think off was,

```
function __cli_perform_completion_once
    if test -n "$results"
        __cli_debug "seems like its already set skipping"
        return 0
    end

    set --global results (__cli_perform_completion)
    __cli_debug "set global results"
    return 0
end
```

and then unsetting it with `__cli_clear_set_globals`, seems like in fish, the order in which completions execute is a stack, so clear set globals will get called after the useful completions at the end

```
complete -c  && __cli_clear_set_globals'
# If this doesn't require order preservation, we don't use the -k flag
complete -c cli -n 'not __cli_requires_order_preservation && __cli_prepare_completions' -f -a '$__cli_comp_results'
# otherwise we use the -k flag
complete -k -c cli -n '__cli_requires_order_preservation && __cli_prepare_completions -f -a '$__cli_comp_results'
```

any thoughts on this?

Reply:
I like that and I don't have any other ideas.
Checking the doc here https://fishshell.com/docs/current/cmds/complete.html
we can read: "Invoking complete multiple times for the same command adds the new definitions on top of any existing completions defined for the command."

So it implies the `complete` commands we register are called in a specific order, which is what we need.

You can go ahead and implement it as you suggest 👍

Reply:
I've implemented this 👍 !